File: src/test/java/practice/common/idempotency/IdempotencyAoPTest.java
package practice.common.idempotency;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.when;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.annotation.Import;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.data.redis.core.ValueOperations;
import org.springframework.http.HttpStatus;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.ResponseEntity;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RestController;


@SpringBootTest
@Import(IdempotencyAoPTest.TestController.class)
class IdempotencyAoPTest {

  @Autowired
  private TestController testController;

  @MockitoBean
  private StringRedisTemplate redisTemplate;

  @Mock
  private ValueOperations<String, String> valueOperations;

  @Test
  void IdempotencyDuplicateCheck2() throws InterruptedException {
    String IDEMPOTENCY_KEY = "test-key";
    Integer THREAD_COUNT = 4;

    when(redisTemplate.opsForValue()).thenReturn(valueOperations);
    when(valueOperations.setIfAbsent("IDEMPOTENCY:" + IDEMPOTENCY_KEY, "EXISTS", 60L, TimeUnit.SECONDS))
        .thenReturn(true)
        .thenReturn(false)
        .thenReturn(false)
        .thenReturn(false);

    ExecutorService executorService = Executors.newFixedThreadPool(THREAD_COUNT);
    CountDownLatch latch = new CountDownLatch(THREAD_COUNT);
    AtomicInteger successCount = new AtomicInteger();
    AtomicInteger conflictCount = new AtomicInteger();

    for (int i = 0; i < THREAD_COUNT; i++) {
      executorService.submit(() -> {
        try {
          HttpStatusCode statusCode = testController.test(IDEMPOTENCY_KEY).getStatusCode();
          if (statusCode == HttpStatus.OK) {
            successCount.incrementAndGet();
          } else if (statusCode == HttpStatus.CONFLICT) {
            conflictCount.incrementAndGet();
          }
        } finally {
          latch.countDown();
        }
      });
    }

    latch.await();
    executorService.shutdown();

    assertEquals(1, successCount.get());
    assertEquals(3, conflictCount.get());

  }


  @RestController
  public static class TestController {

    @GetMapping("/test")
    @IdempotencyCheck("#idempotencyKey")
    public ResponseEntity<?> test(@RequestHeader("Idempotency-Key") String idempotencyKey) {
      try {
        Thread.sleep(1000);
      } catch (InterruptedException e) {
        throw new RuntimeException(e);
      }
      return ResponseEntity.status(HttpStatus.OK).build();
    }
  }
}

File: src/test/kotlin/practice/common/idempotency/IdempotencyAoPTest.kt
package practice.common.idempotency

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.mockito.Mock
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.Import
import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.data.redis.core.ValueOperations
import org.springframework.http.HttpStatus
import org.springframework.http.HttpStatusCode
import org.springframework.http.ResponseEntity
import org.springframework.test.context.bean.override.mockito.MockitoBean
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestHeader
import org.springframework.web.bind.annotation.RestController
import java.util.concurrent.CountDownLatch
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

@SpringBootTest
@Import(IdempotencyAoPTest.TestController::class)
internal class IdempotencyAoPTest {
    @Autowired
    private val testController: TestController? = null

    @MockitoBean
    private val redisTemplate: StringRedisTemplate? = null

    @Mock
    private val valueOperations: ValueOperations<String, String>? = null

    @Test
    @Throws(InterruptedException::class)
    fun IdempotencyDuplicateCheck2() {
        val IDEMPOTENCY_KEY = "test-key"
        val THREAD_COUNT = 4

        Mockito.`when`<ValueOperations<String, String>>(redisTemplate?.opsForValue()).thenReturn(valueOperations)
        Mockito.`when`<Boolean>(
            valueOperations?.setIfAbsent(
                "IDEMPOTENCY:$IDEMPOTENCY_KEY",
                "EXISTS",
                60L,
                TimeUnit.SECONDS
            )
        )
            .thenReturn(true)
            .thenReturn(false)
            .thenReturn(false)
            .thenReturn(false)

        val executorService: ExecutorService = Executors.newFixedThreadPool(THREAD_COUNT)
        val latch: CountDownLatch = CountDownLatch(THREAD_COUNT)
        val successCount: AtomicInteger = AtomicInteger()
        val conflictCount: AtomicInteger = AtomicInteger()

        for (i in 0..<THREAD_COUNT) {
            executorService.submit(Runnable {
                try {
                    val statusCode: HttpStatusCode = testController!!.test(IDEMPOTENCY_KEY).statusCode
                    if (statusCode === HttpStatus.OK) {
                        successCount.incrementAndGet()
                    } else if (statusCode === HttpStatus.CONFLICT) {
                        conflictCount.incrementAndGet()
                    }
                } finally {
                    latch.countDown()
                }
            })
        }

        latch.await()
        executorService.shutdown()

        Assertions.assertEquals(1, successCount.get())
        Assertions.assertEquals(3, conflictCount.get())
    }


    @RestController
    class TestController {
        @GetMapping("/test")
        @IdempotencyCheck("#idempotencyKey")
        fun test(@RequestHeader("Idempotency-Key") idempotencyKey: String?): ResponseEntity<*> {
            try {
                Thread.sleep(1000)
            } catch (e: InterruptedException) {
                throw RuntimeException(e)
            }
            return ResponseEntity.status(HttpStatus.OK).build<Any>()
        }
    }
}
